sql: try harder to trim the filter expression after processing scanNodes.

knz opened the pull request:
Fixes #3473.
Fixes #13351.

cc @andreimatei.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13373)
<!-- Reviewable:end -->

knz committed:
sql: try harder to trim the filter expression after processing scanNodes.

knz commented:
Take your time, it's not too trivial I realize.

(But so far I know only Peter and yourself understand this code.)

RaduBerinde commented:
Great stuff!

---

Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/index_selection.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4mqwMcWTa_i1Rmvw3:-Kc4mqwMcWTa_i1Rmvw4:b-j0um1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L218)):*
> ```Go
> 		if s.filter == parser.DBoolTrue {
> 			s.filter = nil
> 		} else if s.filter == parser.DBoolFalse {
> ```

[nit] easier to read the flow if we do `if it's false { return }` and we don't need the `else`

---

*[pkg/sql/index_selection.go, line 1426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4n-vFCiI0wOTag5Cc:-Kc4n-vFCiI0wOTag5Cd:borobo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1426)):*
> ```Go
> 
> // expandConstraint transforms a potentially complex constraint
> // expression into one or more simple constraint suitable for the
> ```

constraints

---

*[pkg/sql/index_selection.go, line 1464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4py_GTFWPh5E-OYvC:-Kc4py_GTFWPh5E-OYvD:bfbwnyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1464)):*
> ```Go
> 		case parser.LE:
> 			fallthrough
> 		case parser.GE:
> ```

this can just be `case parser.LT, parser.LE,..:`

---

*[pkg/sql/index_selection.go, line 1465 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4nrpMMurSEhb5YsJ0:-Kc4nrpMMurSEhb5YsJ1:b-8mfguc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1465)):*
> ```Go
> 			fallthrough
> 		case parser.GE:
> 			// (a,b,c) < (x,y,z)  -> [a<x]
> ```

This is not true, it implies `a <= x`, e.g. `(1, 2, 3) < (1, 2, 4)`. We need to fix the `GT` and `LT` cases. Also fix the function comment, plus a test would be good given that it wasn't caught. 

---

*[pkg/sql/index_selection.go, line 1522 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4rEztHMOzRoH_D81c:-Kc4rEztHMOzRoH_D81d:b-6tq22n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1522)):*
> ```Go
> // assuming that the expression c on the left (the constraint) is
> // true.
> func applyConstraint(c *parser.ComparisonExpr, t *parser.ComparisonExpr) parser.Expr {
> ```

s/evaluate/simplify?

---

*[pkg/sql/index_selection.go, line 1571 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4pFqG-CytASk5BBKn:-Kc4pFqG-CytASk5BBKo:b-kvu3md) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1571)):*
> ```Go
> }
> 
> // applyConstraintFlat is the common branch of applyConstraint().
> ```

[nit] would still help to explain what the arguments are. Also it would help if the datum and t were together and cOp and cdatum were together.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-Kc4ttP6Z9tLFyBe7wxN:b-ptl4bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
> ```Go
> ) parser.Expr {
> 	switch cOp {
> 	case parser.EQ:
> ```

This code is very tedious, a mistake could easily sneak in (and I doubt that we have tests that cover all the combinations). I have an idea to avoid having 5*5 cases for the comparison ops (NE excluded). Feel free to ignore if though if you don't agree it helps.

We could represent a constraint as a closed interval on the integer axis. Say we pick -100 to represent -infinity, +100 as +infinity. 0 represents the smaller of cdatum and datum (both if they are equal), 10 represents the larger (unused if they are equal).

First set `tx = 0 if datum <= cdatum, or 10 otherwise`.

Then e.g. if `t.Operator=GE`, `t` corresponds to `[tx, +100]`.  If it is LE, `[-100, tx]`. If it is GT, `[tx+1, +100]`.

The other is similar, we just set `cx = 0 if datum >= cdatum or 10 otherwise`.

Then it's simply a matter of detecting if the two intervals are disjoint (return False) or if c is inside t (return True), which are very simple conditions.

---

*[pkg/sql/index_selection.go, line 1774 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4qmLfGur8syLiF24p:-Kc4qmLfGur8syLiF24q:b-lkfztv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1774)):*
> ```Go
> }
> 
> func (v *applyConstraintsVisitor) VisitPost(expr parser.Expr) parser.Expr {
> ```

VisitPost is usually next to VisitPre

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-:-Kc4ru-An-yYJkYXuU3_:b-8o237f)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/index_selection.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4mqwMcWTa_i1Rmvw3:-Kc7r0w3DtyGJ6LBft3H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L218)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] easier to read the flow if we do `if it's false { return }` and we don't need the `else`
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4n-vFCiI0wOTag5Cc:-Kc7rVdKzTJF3Esv3Asn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1426)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

constraints
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4py_GTFWPh5E-OYvC:-Kc7rWA9VhsR5rilUesE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1464)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this can just be `case parser.LT, parser.LE,..:`
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1465 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4nrpMMurSEhb5YsJ0:-Kc8OeKyEX_SbtY1GQjg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1465)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is not true, it implies `a <= x`, e.g. `(1, 2, 3) < (1, 2, 4)`. We need to fix the `GT` and `LT` cases. Also fix the function comment, plus a test would be good given that it wasn't caught. 
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1522 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4rEztHMOzRoH_D81c:-Kc7xNPX6u-8Sgu_K7mv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1522)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

s/evaluate/simplify?
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1571 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4pFqG-CytASk5BBKn:-Kc7xolgbZh189SIQixh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1571)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] would still help to explain what the arguments are. Also it would help if the datum and t were together and cOp and cdatum were together.
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-Kc8GuPrqea1ymZB3vYy:b75h303) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This code is very tedious, a mistake could easily sneak in (and I doubt that we have tests that cover all the combinations). I have an idea to avoid having 5*5 cases for the comparison ops (NE excluded). Feel free to ignore if though if you don't agree it helps.

We could represent a constraint as a closed interval on the integer axis. Say we pick -100 to represent -infinity, +100 as +infinity. 0 represents the smaller of cdatum and datum (both if they are equal), 10 represents the larger (unused if they are equal).

First set `tx = 0 if datum <= cdatum, or 10 otherwise`.

Then e.g. if `t.Operator=GE`, `t` corresponds to `[tx, +100]`.  If it is LE, `[-100, tx]`. If it is GT, `[tx+1, +100]`.

The other is similar, we just set `cx = 0 if datum >= cdatum or 10 otherwise`.

Then it's simply a matter of detecting if the two intervals are disjoint (return False) or if c is inside t (return True), which are very simple conditions.
</blockquote></details>

"Simply a matter of ..." hahaha :trollface: 

Nevertheless the idea was good, so I implemented it.

---

*[pkg/sql/index_selection.go, line 1774 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4qmLfGur8syLiF24p:-Kc7xwhRiJdD5QrTI7I2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1774)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

VisitPost is usually next to VisitPre
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-Kc8YGvd3Fhou4-1s0Fy:b88q79k) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"Simply a matter of ..." hahaha :trollface: 

Nevertheless the idea was good, so I implemented it.
</blockquote></details>

I don't think my proposal quite made it across. An interval would (in the code, *literally*) be represented by two integers instead of datums and inclusive/exclusive flags. The actual datums are not important, just the relationship between them (`cmp`); so we represent them by two fixed values (0 and 10). Exclusivity is implemented by just adding or subtracting 1 from the integer (which is why I chose 0 and 10 and not 0 and 1). Possible values for integers are from the set: -100, -1, 0, +1, 9, 10, 11, 100.

I'll take an example. Say datum < cdatum. and both constraints are LT

tx=0 (as defined above); interval t would be `[-100, -1]`
cx=10; interval c would be `[-100, 9]`

Another example: datum > cdatum, t is LT, c is GE
tx=10; t is `[-100, 9]`
cx=0; c is `[0,+100]`

Once we have these intervals, yes, it's simply a matter of comparing the integers:
 - disjoint:`c.start > t.end || t.start > end`
 - included: `t.start <= c.start && c.start <= t.end`

The simplification is coming from the fact that we reduce the various combinations of cases (nil vs non-nil, inclusive vs exclusive) to a single case: inclusive intervals with comparable values at both ends.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-Kc8s4KD8EqdBtLbPT2K:b4w7ztu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think my proposal quite made it across. An interval would (in the code, *literally*) be represented by two integers instead of datums and inclusive/exclusive flags. The actual datums are not important, just the relationship between them (`cmp`); so we represent them by two fixed values (0 and 10). Exclusivity is implemented by just adding or subtracting 1 from the integer (which is why I chose 0 and 10 and not 0 and 1). Possible values for integers are from the set: -100, -1, 0, +1, 9, 10, 11, 100.

I'll take an example. Say datum < cdatum. and both constraints are LT

tx=0 (as defined above); interval t would be `[-100, -1]`
cx=10; interval c would be `[-100, 9]`

Another example: datum > cdatum, t is LT, c is GE
tx=10; t is `[-100, 9]`
cx=0; c is `[0,+100]`

Once we have these intervals, yes, it's simply a matter of comparing the integers:
 - disjoint:`c.start > t.end || t.start > end`
 - included: `t.start <= c.start && c.start <= t.end`

The simplification is coming from the fact that we reduce the various combinations of cases (nil vs non-nil, inclusive vs exclusive) to a single case: inclusive intervals with comparable values at both ends.
</blockquote></details>

Something along the lines of:

```go
makeInterval := func(op Operator, largerDatum bool) (bool, int, int) {
  x := 0
  if largerDatum {
    x = 10
  }

  switch op {
  case parser.EQ
    return true, x, x
  case parser.LE:
    return true, -100, x
  case parser.LT:
    return true, -100, x-1
  case parser.GE:
    return true, x, 100
  case parser.GT:
    return true, x+1, 100
  case parser.NE:
    return false, x, x
  default:
    return false, 0, 0
  }
}

cmp := datum.Compare(cDatum)
tOk, tStart, tEnd := makeInterval(t.Operator, cmp > 0)
cOk, cStart, cEnd := makeInterval(c.Operator, cmp < 0)
if tOk && cOk {
  if cStart > tEnd || tStart > cEnd {
    return MadeDBool(false)
  }
  if tStart <= cStart && cEnd <= tEnd {
    return MakeDBool(true)
  }
  return t
}
// BOOM: just handled 5*5 combinations of operators!
// If one is NE and the other is an interval, we can still use the intervals
// (hence the NE case above).
if tOk && c.Operator == NE {
  if cStart < tStart || cStart > tEnd {
    return MakeDBool(true)
  }
  return t
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-Kc9uKli0n9FKKHU_0QP:b-ka9e9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Something along the lines of:

```go
makeInterval := func(op Operator, largerDatum bool) (bool, int, int) {
  x := 0
  if largerDatum {
    x = 10
  }

  switch op {
  case parser.EQ
    return true, x, x
  case parser.LE:
    return true, -100, x
  case parser.LT:
    return true, -100, x-1
  case parser.GE:
    return true, x, 100
  case parser.GT:
    return true, x+1, 100
  case parser.NE:
    return false, x, x
  default:
    return false, 0, 0
  }
}

cmp := datum.Compare(cDatum)
tOk, tStart, tEnd := makeInterval(t.Operator, cmp > 0)
cOk, cStart, cEnd := makeInterval(c.Operator, cmp < 0)
if tOk && cOk {
  if cStart > tEnd || tStart > cEnd {
    return MadeDBool(false)
  }
  if tStart <= cStart && cEnd <= tEnd {
    return MakeDBool(true)
  }
  return t
}
// BOOM: just handled 5*5 combinations of operators!
// If one is NE and the other is an interval, we can still use the intervals
// (hence the NE case above).
if tOk && c.Operator == NE {
  if cStart < tStart || cStart > tEnd {
    return MakeDBool(true)
  }
  return t
}
```
</blockquote></details>

Now I understand how it works, thanks for the explanation and the example code. I changed the patch accordingly.
However I am not sure I would have arrived to this myself. Pray, where did you get this idea?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/index_selection.go, line 1576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-Kc4ttP6Z9tLFyBe7wxM:-KcAJkF366GRt4wti2cT:b-ckg59w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ba336bebc8e05a9bc4b07c17e75a0598e168ebc/pkg/sql/index_selection.go#L1576)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Now I understand how it works, thanks for the explanation and the example code. I changed the patch accordingly.
However I am not sure I would have arrived to this myself. Pray, where did you get this idea?
</blockquote></details>

Thanks! I'll take a final look over the entire change by Monday.

Haha, I don't know how I came up with it.. I just hate formulas that have cornercases. I prefer to first apply some transformation if it eliminates checking for special cases. Some examples:
 - whenever I did anything involving computational geometry, I would always choose formulas that don't have special cases like division by zero (e.g. for a line I always used `ax+by+c=0` or `x=x0+t*dx; y=y0+t*dy` instead of `y=mx+b`)
 - if a point was in a bad place for an algorithm (e.g. origin), I'd first translate all points rather than have a special case
 - if I'm doing a dynamic programming formula, I always fill in the 0th row and column with proper values instead of having `if`s in the formula
 - if I'm doing some kind of "fill" on a 2d grid, I surround it with whatever values would prevent the traversal from continuing, rather than checking for the edges.

---

*[pkg/sql/index_selection.go, line 1711 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcAGYOmVAB681A0CLhu:-KcAGYOmVAB681A0CLhv:b-d1963n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1711)):*
> ```Go
> 	//
> 	// We represent a constraint as a closed interval on the integer
> 	// axis. Say we pick -100 to represent -infinity, +100 as
> ```

I wrote down something that might be better for this comment (we can leave out some of the details that can be easily gleamed from the code, and just explain the approach):

```
We map a few interesting points of the datum space to the integer axis as
follows:

                  /- the smaller between datum and cdatum
                  |
 -infinity        |      /- the larger between datum and cdatum
     |            |      |
     |            |      |            /- +infinity
     v            v      v            v
     |------------|------|------------|
   -100           0      10          100

If the datums are equal, they both map to 0.

Constraints become closed intervals on this axis. For example:
a "GT" constraint with the smaller datum becomes [0, 100].  For exclusive
constraints we add or subtract 1 from the datum value: a "GE" constraint with
the smaller datum becomes [1, 100]; a "LE" constraint becomes [-100, -1].

Checking for overlap between the constraints then becomes equivalent to
checking for overlap between the closed intervals (which is easy).
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

knz commented:
@nvanbenschoten for checking whether I'm not doing anything here that conceptually conflicts with #13444.

nvanbenschoten commented:
Reviewed 1 of 6 files at r1, 2 of 4 files at r2.
Review status: 3 of 7 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/sql/index_selection.go, line 210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPMvCjVlW6pGzzEZ_f:-KcPMvCjVlW6pGzzEZ_g:b7pdjci) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L210)):*
> ```Go
> 	if s.filter != nil {
> 		// Constraint propagation may have produced new constant sub-expressions.
> 		// Propagate them.
> ```

nit: Extend `Propagate them.` to something like `Propagate them and check if filter can be applied prematurely.`

---

*[pkg/sql/index_selection.go, line 1448 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPOKN6RXansTXMxeyL:-KcPOKN6RXansTXMxeyM:b-3l8jxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1448)):*
> ```Go
> 			return a
> 		}
> 		vals := *c.Right.(*parser.DTuple)
> ```

After #13444 goes in this will have to be `vals := c.Right.(*parser.DTuple).D`. Same throughout these changes. That said, nothing here conceptually conflicts and neither of the new normalization state flags will need to be considered.

---

*[pkg/sql/index_selection.go, line 1525 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPP2KKdcFnIohCfqnB:-KcPP2KKdcFnIohCfqnC:bbbz33q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1525)):*
> ```Go
> 	switch t := expr.(type) {
> 	case *parser.AndExpr:
> 		if t.Left == parser.DBoolTrue && t.Right == parser.DBoolTrue {
> ```

You passed the Fizzbuzz test! :clap: 

---

*[pkg/sql/index_selection.go, line 1592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPWS8k2Cl7YXNF5xS6:-KcPWS8k2Cl7YXNF5xS7:b-f6u5s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1592)):*
> ```Go
> // [x OP datum] assuming that the expression [x cOp cdatum] (the
> // constraint) is true.
> func applyConstraintFlat(
> ```

nit: it's a little confusing that between `applyConstraint` and `applyConstraintFlat` the "truthy" expression switches places.

---

*[pkg/sql/index_selection.go, line 1711 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcAGYOmVAB681A0CLhu:-KcPVzHIQyX-P9zGAml8:b-i80r2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1711)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I wrote down something that might be better for this comment (we can leave out some of the details that can be easily gleamed from the code, and just explain the approach):

```
We map a few interesting points of the datum space to the integer axis as
follows:

                  /- the smaller between datum and cdatum
                  |
 -infinity        |      /- the larger between datum and cdatum
     |            |      |
     |            |      |            /- +infinity
     v            v      v            v
     |------------|------|------------|
   -100           0      10          100

If the datums are equal, they both map to 0.

Constraints become closed intervals on this axis. For example:
a "GT" constraint with the smaller datum becomes [0, 100].  For exclusive
constraints we add or subtract 1 from the datum value: a "GE" constraint with
the smaller datum becomes [1, 100]; a "LE" constraint becomes [-100, -1].

Checking for overlap between the constraints then becomes equivalent to
checking for overlap between the closed intervals (which is easy).
```

</blockquote></details>

Love the diagrams!

---

*[pkg/sql/index_selection.go, line 1734 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPV37SC7HAqyz3yibO:-KcPV37TK6dhJ90BeGxv:bicmn7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1734)):*
> ```Go
> 	cOk, cStart, cEnd := makeInterval(cOp, cmp < 0)
> 	if tOk && cOk {
> 		if cStart > tEnd || tStart > cEnd {
> ```

Let's pull these two conditions into `disjoint` and `overlapping` boolean (with little diagrams if you're feeling artistic). Then at least `disjoint` can be shared with the `NE/IsNot` case.

---

*[pkg/sql/index_selection.go, line 1743 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPX5BV-YSFD0PE5tvj:-KcPX5BV-YSFD0PE5tvk:bfic9nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1743)):*
> ```Go
> 	}
> 	// BOOM: just handled 5*5 combinations of operators!
> 	// If one is NE and the other is an interval, we can still use the intervals
> ```

nit: qualify which "one is NE", since this matters

---

*[pkg/sql/index_selection.go, line 1744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPU1zv7EsUNe0t2ULn:-KcPU1zwCPYNYdnGxvPO:b-589co7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1744)):*
> ```Go
> 	// BOOM: just handled 5*5 combinations of operators!
> 	// If one is NE and the other is an interval, we can still use the intervals
> 	// (hence the NE case above).
> ```

`s/above/below/`

---

*[pkg/sql/index_selection.go, line 1745 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPXYHoNIeaqkFcJaXm:-KcPXYHoNIeaqkFcJaXn:bpbv48k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1745)):*
> ```Go
> 	// If one is NE and the other is an interval, we can still use the intervals
> 	// (hence the NE case above).
> 	if tOk && (cOp == parser.NE || cOp == parser.IsNot) {
> ```

I must be visualizing this incorrectly, but I feel like this should be `if cOk && (t.Operator == parser.NE || t.Operator == parser.IsNot)`. We're assuming that cOp is true, so if cOp requires that `x != 5`, then we can't simplify `x < 3` to true. Meanwhile, if cOp requires that `x < 3`, then we can simplify `x != 5` to true. Where am I getting tripped up?

---

*[pkg/sql/index_selection.go, line 1755 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPUSRQ8t525KF-5Ygq:-KcPUSRQ8t525KF-5Ygr:b-cagst1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1755)):*
> ```Go
> // makeInterval supports the range comparison in applyConstraintsFlat above.
> // See the comment at the point of call for more details.
> func makeInterval(op parser.ComparisonOperator, largerDatum bool) (bool, int, int) {
> ```

nit: is the boolean return value documented anywhere?

double nit: usually `ok` flags are the last return value

triple nit: `makeInterval` could benefit from a better name. Perhaps `makeComparisonInterval`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/index_selection.go, line 1745 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPXYHoNIeaqkFcJaXm:-KcPaIIZzIHjaeSKnPno:brchlvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1745)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I must be visualizing this incorrectly, but I feel like this should be `if cOk && (t.Operator == parser.NE || t.Operator == parser.IsNot)`. We're assuming that cOp is true, so if cOp requires that `x != 5`, then we can't simplify `x < 3` to true. Meanwhile, if cOp requires that `x < 3`, then we can simplify `x != 5` to true. Where am I getting tripped up?
</blockquote></details>

I definitely got confused as to which is which, maybe we need to give them better names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373)*
<!-- Sent from Reviewable.io -->

knz commented:
Quite a few things to fix with the rebase and the bug y'all have found, but I got there in the end. PTAL!

---

Review status: 3 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPMvCjVlW6pGzzEZ_f:-KcZhzRtaS5BL4Od7965:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L210)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Extend `Propagate them.` to something like `Propagate them and check if filter can be applied prematurely.`
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1448 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPOKN6RXansTXMxeyL:-KcZi0Baejsb8Lc864fP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1448)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

After #13444 goes in this will have to be `vals := c.Right.(*parser.DTuple).D`. Same throughout these changes. That said, nothing here conceptually conflicts and neither of the new normalization state flags will need to be considered.
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPWS8k2Cl7YXNF5xS6:-KcZiPj6_9WSr5aAnv4v:b-l70vjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1592)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: it's a little confusing that between `applyConstraint` and `applyConstraintFlat` the "truthy" expression switches places.
</blockquote></details>

Agreed, fixed.

---

*[pkg/sql/index_selection.go, line 1711 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcAGYOmVAB681A0CLhu:-KcZj-xNPteZzH1Qp_C1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1711)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Love the diagrams!
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1734 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPV37SC7HAqyz3yibO:-KcZjTO4cQ5zPUKVSzVU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1734)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's pull these two conditions into `disjoint` and `overlapping` boolean (with little diagrams if you're feeling artistic). Then at least `disjoint` can be shared with the `NE/IsNot` case.
</blockquote></details>

done

---

*[pkg/sql/index_selection.go, line 1743 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPX5BV-YSFD0PE5tvj:-KcZk0ZhN7sWofP7hAhB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1743)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: qualify which "one is NE", since this matters
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPU1zv7EsUNe0t2ULn:-KcZk0pPhAf83JNxyh8d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1744)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/above/below/`
</blockquote></details>

Done.

---

*[pkg/sql/index_selection.go, line 1745 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPXYHoNIeaqkFcJaXm:-KcZk2IBqPEX26HscmGr:b-hrnv6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1745)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I definitely got confused as to which is which, maybe we need to give them better names.
</blockquote></details>

Done. Also added a test, since this was not caught earlier.

---

*[pkg/sql/index_selection.go, line 1755 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-KcPUSRQ8t525KF-5Ygq:-KcZoxi0vhEPpax3nZPP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63218b1af46611a403cbf3191ce240ff7f0b805/pkg/sql/index_selection.go#L1755)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: is the boolean return value documented anywhere?

double nit: usually `ok` flags are the last return value

triple nit: `makeInterval` could benefit from a better name. Perhaps `makeComparisonInterval`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-:-KcZozucBTbc_AQUWnq_:b-8bzsvm)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 6 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13373#-:-KcZrN-nPDCiXMyESnoP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Ok let's pray this is not going to break everything. TFYR!
